libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,753 @@
+  cls.def(
+      "get_local_edge_id_from_in_edge",
+      [](PropertyGraph& self, GraphTopologyTypes::Edge e) {
+        return self.BuildView<PropertyGraphViews::BiDirectional>()
+            .GetLocalEdgeIDFromInEdge(e);
+      },
+      py::call_guard<py::gil_scoped_release>());
+
+  // GetLocalEdgeID(UndirectedEdgeHandle) -> LocalEdgeID  - local edge ID
+  cls.def(
+      "get_local_edge_id_from_undirected_edge",
+      [](PropertyGraph& self, GraphTopologyTypes::Edge e) {
+        return self.BuildView<PropertyGraphViews::Undirected>()
+            .GetLocalEdgeIDFromUndirectedEdge(e);
+      },
+      py::call_guard<py::gil_scoped_release>());
+
+  // GetEdgePropertyIndex(OutEdgeHandle) -> EdgePropertyIndex  - index into the property table for an edge
+  cls.def(
+      "get_edge_property_index_from_out_edge",
+      &GraphTopology::GetEdgePropertyIndexFromOutEdge);
+
+  // GetEdgePropertyIndex(InEdgeHandle) -> EdgePropertyIndex  - index into the property table for an edge
+  cls.def(
+      "get_edge_property_index_from_in_edge",
+      [](PropertyGraph& self, GraphTopologyTypes::Edge e) {
+        return self.BuildView<PropertyGraphViews::BiDirectional>()
+            .GetEdgePropertyIndexFromInEdge(e);
+      },
+      py::call_guard<py::gil_scoped_release>());
+
+  // GetEdgePropertyIndex(UndirectedEdgeHandle) -> EdgePropertyIndex  - index into the property table for an edge
+  cls.def(
+      "get_edge_property_index_from_undirected_edge",
+      [](PropertyGraph& self, GraphTopologyTypes::Edge e) {
+        return self.BuildView<PropertyGraphViews::Undirected>()
+            .GetEdgePropertyIndexFromUndirectedEdge(e);
+      },
+      py::call_guard<py::gil_scoped_release>());
+
+  // GetEdgePropertyIndex(LocalEdgeID) -> EdgePropertyIndex  - index into the property table for an edge
+
+  // Nodes() -> NodeHandle iterator  - iterator over all graph nodes
+  cls.def("nodes", [](PropertyGraph& self) { return self.Nodes(); });
+
+  // OutEdges() -> OutEdgeHandle iterator  - iterator over all graph out-edges
+  cls.def(
+      "out_edge_ids",
+      py::overload_cast<>(&PropertyGraph::OutEdges, py::const_));
+
+  katana::DefWithNumba<&out_edge_begin>(
+      cls, "_out_edge_begin", katana::numba_only());
+  katana::DefWithNumba<&out_edge_end>(
+      cls, "_out_edge_end", katana::numba_only());
+
+  // OutEdges(NodeHandle)-> OutEdgeHandle iterator - iterator over out-edges for a node
+  cls.def(
+      "out_edge_ids",
+      py::overload_cast<GraphTopologyTypes::Node>(
+          &PropertyGraph::OutEdges, py::const_),
+      py::call_guard<py::gil_scoped_release>());
+
+  // OutEdges(NodeHandle, EdgeEntityTypeID) -> OutEdgeHandle iterator - iterator over out-edges of a type for a node
+  cls.def(
+      "out_edge_ids",
+      [](PropertyGraph& self, GraphTopologyTypes::Node n,
+         const EntityType& ty) {
+        return self.BuildView<PropertyGraphViews::EdgeTypeAwareBiDir>()
```

Comment:
so does this mean we can't keep around a `view` object in python?

Reply:
Access views are abstracted away in Python yes. You just get the universal view which allows everything. We may have to allow explicit instances of views for numba compiled algorithm implementations, but that should be a special case not the normal usage.